### PR TITLE
Batch insert return id

### LINKED ISSUE #5040: Batch insert via mybatis,useGeneratedKeys not work
### Which version of ShardingSphere did you use?

* sharding-jdbc-core
  4.0.1

* mybatis-spring-boot-starter

  2.1.2

* mysql-connector-java

  8.0.19

* spring-boot

  2.2.2.RELEASE

### Which project did you use? Sharding-JDBC or Sharding-Proxy?

Sharding-JDBC

### Expected behavior

Using mybatis to batch insert records, write back the auto generated keys(id) to entity's key filed. 

### Actual behavior

Entity's key field is still null.

### Reason analyze (If you can)

after invoke org.apache.shardingsphere.shardingjdbc.jdbc.core.statement.ShardingPreparedStatement#shard
SQLRouteResult.generatedKey is null

### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.

* sql to create table:

``` sql
CREATE TABLE `t_user0` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `name` varchar(255) COLLATE utf8_bin DEFAULT NULL,
  `sharding_key` int(255) NOT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=1 DEFAULT CHARSET=utf8 COLLATE=utf8_bin

CREATE TABLE `t_user1` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `name` varchar(255) COLLATE utf8_bin DEFAULT NULL,
  `sharding_key` int(255) NOT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=100000001 DEFAULT CHARSET=utf8 COLLATE=utf8_bin

CREATE TABLE `t_user2` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `name` varchar(255) COLLATE utf8_bin DEFAULT NULL,
  `sharding_key` int(255) NOT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=200000001 DEFAULT CHARSET=utf8 COLLATE=utf8_bin

CREATE TABLE `t_user3` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `name` varchar(255) COLLATE utf8_bin DEFAULT NULL,
  `sharding_key` int(255) NOT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=300000001 DEFAULT CHARSET=utf8 COLLATE=utf8_bin
```

* sharding config

``` yaml
dataSources:
  ds_0: !!com.zaxxer.hikari.HikariDataSource
    driverClassName: com.mysql.cj.jdbc.Driver
    jdbcUrl: jdbc:mysql://127.0.0.1:3306/test?serverTimezone=GMT%2B8&useUnicode=true&characterEncoding=utf-8
    username: root
    password: root
    maximumPoolSize: 20

shardingRule:
  tables:
    t_user:
      actualDataNodes: ds_0.t_user${0..3}
      databaseStrategy:
        none:
      tableStrategy:
        inline:
          algorithmExpression: t_user${sharding_key % 4}
          shardingColumn: sharding_key
  bindingTables:
    - t_user
  defaultDataSourceName: ds_0

props:
  sql.show: true
```

* UserDO.java

``` java
@Data
@NoArgsConstructor
public class UserDO implements Serializable {
    private Integer id;

    private String name;

    private Integer shardingKey;

    private static final long serialVersionUID = 1L;

	public UserDO(String name, Integer shardingKey) {
		this.name = name;
		this.shardingKey = shardingKey;
	}
}
```

* UserMapper.java

``` java
@Mapper
public interface UserMapper {
	int batchInsert(List<UserDO> list);
}
```

* UserMapper.xml

``` xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE mapper PUBLIC "-//mybatis.org//DTD Mapper 3.0//EN" "http://mybatis.org/dtd/mybatis-3-mapper.dtd">
<mapper namespace="com.jack.test.sharding.jdbc.dal.mapper.UserMapper">
  <resultMap id="BaseResultMap" type="com.jack.test.sharding.jdbc.dal.dataobject.UserDO">
    <id column="id" jdbcType="INTEGER" property="id" />
    <result column="name" jdbcType="VARCHAR" property="name" />
    <result column="sharding_key" jdbcType="INTEGER" property="shardingKey" />
  </resultMap>
  <sql id="Base_Column_List">
    id, `name`, sharding_key
  </sql>
    <insert id="batchInsert" keyColumn="id" keyProperty="id" useGeneratedKeys="true">
        insert into t_user (`name`, sharding_key)
        values
        <foreach collection="list" item="item" index="index" separator=",">
            (
            #{item.name,jdbcType=VARCHAR}, #{item.shardingKey,jdbcType=INTEGER}
            )
        </foreach>
    </insert>
</mapper>
```

* DataSourceConfig.java

``` java
@Configuration
public class DataSourceConfig {
	@Bean
	public DataSource dataSource() {
		try {
			String yamlStr = org.apache.commons.io.IOUtils.toString(new ClassPathResource("application-sharding-databases.yml").getInputStream(), StandardCharsets.UTF_8);
			DataSource dataSource = YamlShardingDataSourceFactory.createDataSource(yamlStr.getBytes());
			return dataSource;
		} catch (Exception e) {
			throw new RuntimeException(e);
		}
	}
}
```

* ShardingTest.java

``` java
	@Test
	void batchInsert() {
		List<UserDO> users = IntStream.range(1,101).boxed().map(i->new UserDO("name"+i,i)).collect(Collectors.toList());
		userMapper.batchInsert(users);
		for (UserDO user : users) {
			assertNotNull(user.getId());
		}
	}

```

### PR DESCRIPTION
**Pull request by zjcnb**
Fixes #5040

batch insert return ids

### CODE REVIEW

**Review by tristaZero — CHANGES_REQUESTED** [review]
A UT is necessary

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/loader/ColumnMetaDataLoader.java [hunk 1]
```diff
@@ -76,15 +75,13 @@
                 }
             }
         }
-        try (Statement statement = connection.createStatement(); ResultSet resultSet = statement.executeQuery(generateEmptyResultSQL(tableNamePattern, databaseType))) {
-            for (String each : columnNames) {
-                isCaseSensitives.add(resultSet.getMetaData().isCaseSensitive(resultSet.findColumn(each)));
+        try (ResultSet resultSet = connection.createStatement().executeQuery(generateEmptyResultSQL(tableNamePattern, databaseType))) {
```
Comment (tristaZero) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/loader/ColumnMetaDataLoader.java:78]: Do you think the statement from `createStatement()` still needs to be closed?

**Review by tristaZero — CHANGES_REQUESTED** [review]

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/loader/ColumnMetaDataLoader.java (hunk 1 above)
Comment (zjcnb) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/loader/ColumnMetaDataLoader.java:78]: yes, i think it ‘s still needs closed

File: shardingsphere-jdbc/shardingsphere-jdbc-core/src/test/java/org/apache/shardingsphere/driver/jdbc/core/statement/EncryptPreparedStatementTest.java [hunk 2]
```diff
@@ -84,7 +84,7 @@ public void assertInsertWithBatchExecute() throws SQLException {
     
     @Test
     public void assertInsertWithExecuteWithGeneratedKey() throws SQLException {
-        try (PreparedStatement statement = getEncryptConnection().prepareStatement(INSERT_GENERATED_KEY_SQL, Statement.RETURN_GENERATED_KEYS)) {
```
Comment (strongduanmu) [inline shardingsphere-jdbc/shardingsphere-jdbc-core/src/test/java/org/apache/shardingsphere/driver/jdbc/core/statement/EncryptPreparedStatementTest.java:87]: @zjcnb Why modify these two unit tests?

Comment (zjcnb) [reply shardingsphere-jdbc/shardingsphere-jdbc-core/src/test/java/org/apache/shardingsphere/driver/jdbc/core/statement/EncryptPreparedStatementTest.java:87]: When I ran the unit tests, these two didn't passed, but now they pssed, so I'll restore them

**Review by strongduanmu — APPROVED** [review]

**Review by tristaZero — APPROVED** [review]

**Comment by tristaZero** [comment]
@zjcnb @strongduanmu Nice work, guys.

**Pull request merged**